### PR TITLE
Support transient utilization attributes

### PR DESCRIPTION
**Pull request by clumens**
This seems suspiciously easy.

Basically, you use attrd_updater -z when creating an attribute to specify it's a utilization attribute (just like in crm_attribute).  After that, you don't need to use -z when updating or deleting - it'll just know which block the attribute is in.  The CIB ends up with something like this in it:

```
      <transient_attributes id="1">
        <instance_attributes id="status-1">
          <nvpair id="status-1-.feature-set" name="#feature-set" value="3.16.2"/>
          <nvpair id="status-1-master-stateful" name="master-stateful" value="5"/>
          <nvpair id="status-1-last-failure-httpd-bundle-podman-0.start_0" name="last-failure-httpd-bundle-podman-0#start_0" value="1668202467"/>
          <nvpair id="status-1-fail-count-httpd-bundle-podman-0.start_0" name="fail-count-httpd-bundle-podman-0#start_0" value="INFINITY"/>
          <nvpair id="status-1-last-failure-httpd-bundle-podman-1.start_0" name="last-failure-httpd-bundle-podman-1#start_0" value="1668202472"/>
          <nvpair id="status-1-fail-count-httpd-bundle-podman-1.start_0" name="fail-count-httpd-bundle-podman-1#start_0" value="INFINITY"/>
        </instance_attributes>
        <utilization id="status-1">
          <nvpair id="status-1-ABC" name="ABC" value="123"/>
        </utilization>
      </transient_attributes>
    </node_state>
```

Things that still need to be tested:

* [x] Is it possible to have two attributes with the same name where one is a utilization attr and the other is an instance attr?
* [x] I haven't checked that the value of a transient utilization attr actually gets included when calculating utilization.
* [x] I haven't checked that a multi-attr message works here, but I don't see why it wouldn't.  As an aside - it'd be great to have a way to test this that doesn't involve hacking some code in here and there every time.  Some sort of test program, or compile-time #define, or something would be great.
* [x] I haven't tested this with sync points, but again I don't see why it wouldn't work.  Nothing has changed there.

### CODE REVIEW

**Comment by clumens** [comment]
>     * [ ]  Is it possible to have two attributes with the same name where one is a utilization attr and the other is an instance attr?

This is a little bit weird.

With attrd_updater... first run `attrd_updater -z -n cpu -U 1` and you get a utilization attribute created.  If you then try to do `attrd_updater -n cpu -U 2`, it automatically updates the utilization attribute to have a value of 2 instead of creating a regular transient attribute.

With crm_attribute... first run `crm_attribute --node rhel8-cluster-1 --name mem --update 1024 -z` and you get a utilization attribute created.  If you then try to do `crm_attribute --node rhel8-cluster-1 --name mem --update 2048`, it creates a new regular attribute and preserves the value of the transient attribute.

I don't really like the inconsistency in behavior there, though I haven't tracked down exactly why it's happening.  @kgaillot any suggestions over what should be done here - keep this behavior or change one of the two programs?

**Comment by clumens** [comment]
Following up on this, the difference is that crm_attribute and attrd_updater use completely different code paths to store/lookup/update attributes.

> With attrd_updater... first run `attrd_updater -z -n cpu -U 1` and you get a utilization attribute created. If you then try to do `attrd_updater -n cpu -U 2`, it automatically updates the utilization attribute to have a value of 2 instead of creating a regular transient attribute.

attrd_updater uses attrd for everything.  attrd uses a hash table keyed by attribute name to store attributes (see attrd_populate_attribute).  The set is not taken into account anywhere.  Once an attribute name is used in one set, any subsequent call to any attrd IPC function will work on that attribute regardless of whether you give it another set name or even no set name at all.

> With crm_attribute... first run `crm_attribute --node rhel8-cluster-1 --name mem --update 1024 -z` and you get a utilization attribute created. If you then try to do `crm_attribute --node rhel8-cluster-1 --name mem --update 2048`, it creates a new regular attribute and preserves the value of the transient attribute.

crm_attribute does not use attrd by default for updates.  It instead constructs a big xpath query (see the find_attr function) and uses that to look up an attribute in the CIB.  The xpath query takes the set_type into account, which means an attribute name can exist in the instance_attributes, meta_attributes, and utilization sets all at the same time.

Making attrd_updater work the same way would require reworking the hash table to involve both the attribute name and the attribute set, which seems like a fair amount of work.  On the other hand, it sure would be nice to get these two programs working as similarly as possible in the hopes of one day merging them.

File: include/crm/common/ipc_attrd_internal.h
Comment (kgaillot) [inline include/crm/common/ipc_attrd_internal.h:150]: I'm thinking this could just be a flag in pcmk__node_attr_opts

**Review by kgaillot — COMMENTED** [review]
Nice :)

> Is it possible to have two attributes with the same name where one is a utilization attr and the other is an instance attr?

Pacemaker generally assumes that the user won't do anything nonsensical, which of course is not a good idea. :)

Since the CIB can be modified arbitrarily, it probably would be a good idea to handle that in some reasonable manner. The current tool behavior is acceptable, I think. Ideally, both would behave by default like attrd_updater currently behaves, but allow the user to override it the way crm_attribute currently behaves. But that can be left for when they're eventually consolidated.

> I haven't checked that the value of a transient utilization attr actually gets included when calculating utilization.

Yep, that's the missing piece. :) unpack_nodes() unpacks permanent utilization attributes (i.e. under \<node\>). add_node_attrs() unpacks transient_attributes (i.e. under \<node_state\>), so it will need to do something similar.

> I haven't checked that a multi-attr message works here, but I don't see why it wouldn't. As an aside - it'd be great to have a way to test this that doesn't involve hacking some code in here and there every time. Some sort of test program, or compile-time #define, or something would be great.

That would be adding an option to attrd_updater and/or crm_attribute to set a list.

File: include/crm/common/ipc_attrd_internal.h
Comment (clumens) [reply include/crm/common/ipc_attrd_internal.h:150]: Yeah, that works as long as we won't want to set them into another block in the future.

**Comment by clumens** [comment]
> Since the CIB can be modified arbitrarily, it probably would be a good idea to handle that in some reasonable manner. The current tool behavior is acceptable, I think. Ideally, both would behave by default like attrd_updater currently behaves, but allow the user to override it the way crm_attribute currently behaves. But that can be left for when they're eventually consolidated.

Right, at least now I know why things are the way they are.

> Yep, that's the missing piece. :) unpack_nodes() unpacks permanent utilization attributes (i.e. under <node>). add_node_attrs() unpacks transient_attributes (i.e. under <node_state>), so it will need to do something similar.

Okay, I see where this is happening.

> That would be adding an option to attrd_updater and/or crm_attribute to set a list.

And I think we previously decided it was very difficult to come up with a command line where this would make sense.  One idea might be to allow one of these tools to take an XML string on stdin or the command line, like other tools do.

File: include/crm/common/ipc_attrd_internal.h
Comment (kgaillot) [reply include/crm/common/ipc_attrd_internal.h:150]: I don't expect any others, and it's internal API so at least easy to change

**Comment by kgaillot** [comment]
> > That would be adding an option to attrd_updater and/or crm_attribute to set a list.
> 
> And I think we previously decided it was very difficult to come up with a command line where this would make sense. One idea might be to allow one of these tools to take an XML string on stdin or the command line, like other tools do.

Not necessarily difficult, just completely different from the current interface. Maybe something like: if --name is not given, then --update (or --update-both) can be specified multiple times with NAME=VALUE like:

    attrd_updater --node=node1 --update connected=1 --update source=node2

Alternatively to follow the existing interface more closely we could just allow --name and --update/--update-both to be specified multiple times, requiring that they match up like:

    attrd_updater --node=node1 -n connected -U 1 -n source -U node2

File: include/crm/common/ipc_attrd_internal.h
Comment (clumens) [reply include/crm/common/ipc_attrd_internal.h:150]: Fixed.

File: daemons/attrd/attrd_cib.c
Comment (kgaillot) [inline daemons/attrd/attrd_cib.c:140]: probably worth an else error

File: daemons/attrd/pacemaker-attrd.h
Comment (kgaillot) [inline daemons/attrd/pacemaker-attrd.h:123]: I've always found the "set" terminology obscure, and now that we're adding block it looks even more so. I'm thinking something like block_type+block_id, set_type+set_id, or parent_type+parent_id.

File: tools/attrd_updater.c
Comment (kgaillot) [inline tools/attrd_updater.c:218]: The utilization terminology is a little confusing. This could be considered a resource utilization attribute, but we refer to it elsewhere as a node utilization attribute, so we should stick with that. It's worth mentioning that this will only be used if the attribute needs to be created, otherwise its existing type will continue to be used regardless. Also, bump the feature set.

FYI, I eventually want to use "capacity" for nodes and keep "utilization" for resources, but that's a bigger project for the future.

**Review by kgaillot — COMMENTED** [review]
Love it, very clean

**Comment by kgaillot** [comment]
It sounds like we need the new commits for 2.1. Also ++ support is critical since that's how fail counts are increased -- I'm not sure how (if?) that is working currently.

**Comment by clumens** [comment]
> It sounds like we need the new commits for 2.1. Also ++ support is critical since that's how fail counts are increased -- I'm not sure how (if?) that is working currently.

I've pushed #2948 with these fixes and removed them from this PR.  Let's hold off on merging this one until after that PR is merged onto 2.1 and then onto main as well.  After that's done I've got another followup that will only apply to main.  Then we can come back and address this PR again.  In the meantime, I'll address the other comments here.

File: tools/attrd_updater.c
Comment (clumens) [reply tools/attrd_updater.c:218]: Fixed, though I'm not especially happy with the new description.

File: daemons/attrd/pacemaker-attrd.h
Comment (clumens) [reply daemons/attrd/pacemaker-attrd.h:123]: I don't think we can change the names of existing attributes in attrd messages, right?  Even though it's all internal, it'll cause problems for mixed version clusters.

With that in mind, we already have a->set = PCMK__XA_ATTR_SET = "attr_set".  I like keeping the structure name and the attribute name at least somewhat related so you don't have to think extra hard.  So maybe rename a->set to a->set_id, rename a->block to a->set_type, and renamed PCMK__XA_ATTR_BLOCK/"attr_block" to PCMK__XA_ATTR_SET_TYPE/"attr_set_type"?

Comment (kgaillot) [reply daemons/attrd/pacemaker-attrd.h:123]: > I don't think we can change the names of existing attributes in attrd messages, right? Even though it's all internal, it'll cause problems for mixed version clusters.

Correct. It's never user-visible so we don't have to care about changing them. It's just the struct members so we get less confused :)

> 
> With that in mind, we already have a->set = PCMK__XA_ATTR_SET = "attr_set". I like keeping the structure name and the attribute name at least somewhat related so you don't have to think extra hard. So maybe rename a->set to a->set_id, rename a->block to a->set_type, and renamed PCMK__XA_ATTR_BLOCK/"attr_block" to PCMK__XA_ATTR_SET_TYPE/"attr_set_type"?

Sounds good

File: daemons/attrd/attrd_cib.c
Comment (clumens) [reply daemons/attrd/attrd_cib.c:140]: Added.

**Comment by clumens** [comment]
This is the most recent version of the patches, but I still need to go through and do more thorough testing again (especially in light of all the other attrd IPC stuff I changed recently).  So, not quite ready to merge.

**Comment by clumens** [comment]
Added documentation and retested everything.  If there's nothing else to do, I think this is good to go.

File: tools/attrd_updater.c
Comment (kgaillot) [inline tools/attrd_updater.c:223]: Will crm_attribute --utilization --lifetime=reboot automatically work now?

File: lib/pengine/unpack.c
Comment (kgaillot) [inline lib/pengine/unpack.c:4239]: We should add a scheduler regression test for this. It should be sufficient to take one of the existing utilization tests and make one of the node attributes transient.

**Review by kgaillot — COMMENTED** [review]
looks good

File: tools/attrd_updater.c
Comment (clumens) [reply tools/attrd_updater.c:223]: It doesn't look like --lifetime= does anything.  nodes/forever sets the pcmk__node_attr_perm flag, and status/reboot clears it.  But it's never actually used anywhere.

Comment (kgaillot) [reply tools/attrd_updater.c:223]: Not attrd_updater, crm_attribute

Comment (clumens) [reply tools/attrd_updater.c:223]:
```
[root@rhel8-cluster-1 pacemaker]# crm_attribute --utilization --lifetime=reboot --name BBB --update 222
[root@rhel8-cluster-1 pacemaker]# cibadmin -Q | less
...
      <transient_attributes id="1">
        <instance_attributes id="status-1">
          <nvpair id="status-1-BBB" name="BBB" value="222"/>
        </instance_attributes>
        <utilization id="status-1">
          <nvpair id="status-1-AAA" name="AAA" value="123"/>
        </utilization>
      </transient_attributes>
```

So close.

File: lib/pengine/unpack.c
Comment (clumens) [reply lib/pengine/unpack.c:4239]: Added.

File: tools/attrd_updater.c
Comment (clumens) [reply tools/attrd_updater.c:223]: Looks like something's still wrong in crm_attribute:
```
(pcmk__send_ipc_request)        trace: ipc-sent   <create_attrd_op t="attrd" src="crm_attribute" attr_name="BBB" attr_set_type="instance_attributes" task="update" attr_value="222" attr_host="rhel8-cluster-1" attr_is_remote="0" attr_is_private="0"/>
```

vs when run from attrd_updater:
```
(pcmk__send_ipc_request)        trace: ipc-sent   <create_attrd_op t="attrd" src="attrd_updater" attr_name="AAA" attr_set_type="utilization" task="update" attr_value="123" attr_is_remote="0" attr_is_private="0"/>
```

Note the attr_set_type attribute.  I can fix that as part of this PR, or we can do a second one.  Whichever.

Comment (kgaillot) [reply tools/attrd_updater.c:223]: Let's do it with this for completeness

Comment (clumens) [reply tools/attrd_updater.c:223]: No problem.

**Review by kgaillot — APPROVED** [review]

**Pull request merged**
